src/Microsoft.Management.UI.Internal/ManagementList/FilterCore/IDeepCloneable.cs

```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Management.UI.Internal
+{
+    /// <summary>
+    /// Defines a generalized method for creating a deep copy of an instance.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.MSInternal", "CA903:InternalNamespaceShouldNotContainPublicTypes")]
+    public interface IDeepCloneable
+    {
+        /// <summary>
+        /// Creates a deep copy of the current instance.
+        /// </summary>
+        /// <returns>A new object that is a deep copy of the current instance.</returns>
+        object DeepClone();
+    }
```

Comment:
I'd prefer strong typed interface. It is more safe.
```suggestion
    public interface IDeepCloneable<TSelf>
        where TSelf : class
    {
        /// <summary>
        /// Creates a deep copy of the current instance.
        /// </summary>
        /// <returns>A new object that is a deep copy of the current instance.</returns>
        TSelf DeepClone();
    }
```

Reply:
I wanted to do that initially, but then I can no longer check if a type derived from TSelf implements the interface. Example:
```cs
public abstract class FilterRule : IDeepCloneable<FilterRule> {}
public class SomeRule<T> : FilterRule {}
public class Foo<T> {
    public T TypedValue { get; set; }
    public object Value { get; set; }
    public bool IsTypedValueDeepCloneable => typeof(IDeepCloneable<T>).IsAssignableFrom(typeof(T));
    public bool IsValueDeepCloneable => Value is IDeepCloneable<T>;
}
var foo = new Foo<SomeRule<int>>();
foo.TypedValue = new SomeRule<int>();
foo.Value = new SomeRule<int>();
var oops1 = foo.IsTypedValueDeepCloneable; // wrongly returns false.
var oops2 = foo.IsValueDeepCloneable; // wrongly returns false.
```

Reply:
Strong typed interface assumes that we have to add it to every class and override DeepClone() in every class - such specific DeepClone() should use right constructor of the class.

Reply:
FilterRule implements DeepClone() and instantiates the derived classes via copy constructor. The copy constructors are needed anyway, because initialization has to be done slightly different than the default ctor does. (what BinaryFormatter did via `[OnDeserialized]`).

Reply:
Here is a [more comprehensive demo](https://gist.github.com/mawosoft/52055e772a88cdf516b7e13b3d394565) of the problems with the strongly typed interface, be it implemented on the base class or on every derived class. Neither works properly for all use cases.

Reply:
I mean follow:
```c#
Console.WriteLine("Hello, World!");
var step1 = new DerivedRuleA();
var step2 = step1.DeepClone();
Console.WriteLine($"{step2}");
Console.WriteLine("");

var step3 = new DerivedRuleB();
var step4 = step3.DeepClone();
Console.WriteLine($"{step4}");
Console.WriteLine("");

DerivedRuleA step5 = new DerivedRuleB();
DerivedRuleA step6 = step5.DeepClone();
Console.WriteLine($"{step6}");
Console.WriteLine("");

public interface IDeepCloneable<TSelf>
{
    TSelf DeepClone();
}

public abstract class FilterRuleA : IDeepCloneable<FilterRuleA>
{
    public abstract FilterRuleA DeepClone();
}

public class DerivedRuleA : FilterRuleA
{
    public DerivedRuleA() => Console.WriteLine(">Called DerivedRuleA()");
    public DerivedRuleA(DerivedRuleA s) => Console.WriteLine(">Called DerivedRuleA(DerivedRuleA s)");
    public override DerivedRuleA DeepClone()
    {
        Console.WriteLine(">Called DerivedRuleA DeepClone()");
        return new DerivedRuleA(this);
    }
}

public class DerivedRuleB : DerivedRuleA
{
    public DerivedRuleB() => Console.WriteLine(">Called DerivedRuleB()");
    public DerivedRuleB(DerivedRuleB s) => Console.WriteLine(">Called DerivedRuleB(DerivedRuleB s)");
    public override DerivedRuleB DeepClone()
    {
        Console.WriteLine(">Called DerivedRuleB DeepClone()");
        return new DerivedRuleB(this);
    }
}
```

Reply:
I understand what you want but it doesn't work in ~~real life~~ the scenario we are dealing with.

We have cloneable types that contain other cloneable types as properties, the exact types of which are not precisely known. The generic type of such a property may be instantiated as a base class like `FilterRule` or `ValidatingValueBase`, some derived class like `TextEqualsFilterRule` or `ValidatingValue`, or it may be some other type that doesn't implement our clone interface, like `DateTime` or `string`.

We need to check for the interface being available to decide whether to (deep)clone or do a simple copy - which is impossible with your suggestion.

Reply:
Perhaps I skip something. I believe all the work is about `static FilterRule DeepCopy(this FilterRule rule)` method using BinaryFormatter. So I don't understand "or it may be some other type that doesn't implement our clone interface, like DateTime or string". Can you clarify?

Reply:
The reason why the original author chose to use Binaryformatter instead of doing the work himself (presumably) is the complexity of the involved generic and non-generic types and how they are nested. If you recall, there also was an attempt at replacing Binaryformatter two years ago, which failed because it did not fully take that complexity into account.

`FilterRule` is the base class of a hierarchy of ~20 classes, nested up to six levels. Some of them are generic. Their generic properties are of type `ValidatingValue<T>` and `ValidatingSelectorValue<T>`.

That type `T` can be an "incoming" type corresponding to the columns of the gridview, which I referred to with "some other type like string etc".

But `T` can also be another class derived from `FilterRule`, so you get multiple nested generics, some of which are generated dynamically at runtime via `MakeGenericType()`.

Reply:
All the copying work is done by the new copy constructor (you have already created these constructors), which is called by a virtual method from the interface.  Therefore, there should be no problems that you are talking about.